core/src/main/java/org/apache/accumulo/core/iterators/user/ScanYieldException.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.core.iterators.user;
+
+import org.apache.accumulo.core.data.Key;
+
+/**
+ * This exception can be thrown at from a next or seek call on an iterator to allow other scans to get time slots. This mechanism is intended to avoid a set of
+ * scans to dominate all of the scan slots (readahead threads) and starve other scans out.
+ */
+public class ScanYieldException extends RuntimeException {
+  private final Key position;
+
+  public ScanYieldException(Key position) {
```

Comment:
the seek() would pass in the key at which it left off (somewhere past the start of the range)
the next() would also pass in the key at which it left which should not be the last key returned by the iterator.

It appears that I have not adequately described the process here.  Here is a scenario that might help understanding:

Lets say we have a tablet with keys a, b, c, ... z.
Someone writes an iterator which it to only return keys b, and z.
Seek is called with a range of (-inf, inf).
The iterator starts scanning and finds key b and returns normally.
The tablet calls hasTop-> true and calls getTopKey() which returns key b.
The tablet then calls next.
The Iterator next call continues scanning and gets to key m and wants to return control to the tablet so it throws the ScanYieldException setting the key to m
The tablet comes back sometime later and calls seek with a range of (m, inf).
The iterator seek call continues scanning the underlying keys and gets to key u and wants to return control to the tablet so it throws a ScanYieldException setting the key to u
The tablet comes back sometime later and calls seek with a range of (u, inf)
The iterator seek calls continues scanning and gets to key z and returns normally.
The tablet calls hasTop-> true and calls getTopKey() which returns key z.
The tablet calls next()
Next returns immediately as there are no more keys
The table calls hasTop -> false and the scan is complete.